### ISSUE TITLE
Improve `imulc` Performance

### ISSUE THREAD

**Issue by nlordell** [nlordell/ethnum-rs#21]
Checked signed integer multiplication (`imulc` and `I256::checked_mul`) use **division** to verify overflows. This is, I _think_, quite inefficient and unnecessary.

This issue captures the work to:
- [ ] Benched checked signed multiplication
- [ ] Investigate and implement possible alternative algorithms

One idea would be to multiply absolutes and then verify the sign afterwards. I'm not sure if this will work for `I256::overflowing_mul` (in that the wrapping produces different results), but it might be worth a try!

**Comment by NCGThompson** [comment]
tl;dr: nlordell's idea of verifying the sign afterwards should work and his concern of wrapping problems is unfounded. There is a tiny caveat with `I256::MIN`.

The advantage of two's complement is that the values wrap around correctly, and are congruent modulo 2<sup>bits</sup> to their binary representation reinterpreted as an unsigned integer. This means that we can do modulo arithmetic on congruence classes and apply it to signed and unsigned integers simultaneously.

`imulc` is given two integers, *a&#x305;*<sub>256</sub> and *b&#x305;*<sub>256</sub>. One of our outputs, *c&#x305;*<sub>256</sub>, is the modular product of our two inputs. The `bool` output *overflow*, on the other hand, is partial to whether or not we are interpreting the values as signed or unsigned.

Let's define functions that interpret the congruence classes as regular integers. signed(x) interprets x as a signed integer while unsigned(x) interprets x as an unsigned integer. So for example signed(-1&#x305;0<sub>256</sub>) = -10 while unsigned(-1&#x305;0<sub>256</sub>) = 2<sup>256</sup> - 10.

*overflow* should return positive iff signed(*a&#x305;*<sub>256</sub>) * signed(*b&#x305;*<sub>256</sub>) &ne; signed(*c&#x305;*<sub>256</sub>). Equivalently, *overflow* should return false iff signed(`I256::MIN`) &leq; signed(*a&#x305;*<sub>256</sub>) * signed(*b&#x305;*<sub>256</sub>) &leq; signed(`I256::MAX`).

> One idea would be to multiply absolutes and then verify the sign afterwards.

That makes sense. If `umulc(...).1` of the absolute values returns `true`, then we know *overflow* is true. I'm not going to bother to explain that with math symbols. If `umulc` didn't overflow and signed(product of absolutes) is non-negative, then we know *overflow* is false. If the product is negative, then *overflow* is true unless the product is `I256::MIN` and exactly one of our inputs is negative.

>  I'm not sure if this will work for I256::overflowing_mul (in that the wrapping produces different results)

We wan't to be able to "undo" the absolute value at the end by negating the product if exactly one of the inputs was negative. So the question is whether or not *c&#x305;*<sub>256</sub> &equiv; 0&#x305;<sub>256</sub> - (0&#x305;<sub>256</sub> - *a&#x305;*<sub>256</sub>) * *b&#x305;*<sub>256</sub> and *c&#x305;*<sub>256</sub> &equiv; (0&#x305;<sub>256</sub> - *a&#x305;*<sub>256</sub>) * (0&#x305;<sub>256</sub> - *b&#x305;*<sub>256</sub>)

With modulo arithmetic, congruence is preserved with addition, subtraction, and multiplication. Therefor we can rewrite the question more simply: "Does *a* * *b* = -(-*a* * *b*) and *a* * *b* = -*a* * - *b*?" This is clearly true.

**Comment by NCGThompson** [comment]
This is how llvm does it. This function is called on Aarch64 for i128 mul. https://github.com/llvm/llvm-project/blob/main/compiler-rt/lib/builtins/int_mulo_impl.inc
